Add support for user-supplied palette + indexed pixel input

nurpax opened the pull request:
Quantizing (palette computation + remapping RGB to indexed) is a slow
process.  If the gif encoder inputs already happen to be indexed
(which it often is when say encoding pixel art into gif anims), this
whole process can be skipped.

This patch allows specifying a palette once on GIFEncoder creation and
makes it possible to pass indexed inputs to GIFEncoder.addFrame.

This is somewhat of a prototype.  Can clean this up if it looks like
this has a chance to be merged.

nurpax committed:
Add support for user-supplied palette + indexed pixel input

Quantizing (palette computation + remapping RGB to indexed) is a slow
process.  If the gif encoder inputs already happen to be indexed
(which it often is when say encoding pixel art into gif anims), this
whole process can be skipped.

This patch allows specifying a palette once on GIFEncoder creation and
makes it possible to pass indexed inputs to GIFEncoder.addFrame.

This is somewhat of a prototype.  Can clean this up if it looks like
this has a chance to be merged.

twolfson commented:
Sorry for the slow reply, was on vacation. Will review by the end of next week. Seems fine at a glance

twolfson commented:
Looks good to me. What were the ideas you had for clean-up?

I think before I land I'll readjust everything so it's all `var's` still for consistency as well as add documentation but it sounds like you had other ideas

nurpax commented:
Now indexed vs rgb pixel data is heuristically detected based on pixels array length.  A more orthogonal implementation would support all these 3 options:

1. RGB pixels, compute palette, quantize
2. RGB pixels + input palette, quantize to given palette (not supported in my patch)
3. indexed pixels + input palette, basically just pass through everything (this is what I added)

Option 2 would be nice to have as the API would be a bit more orthogonal.

But #3 is what I was after I needed in https://github.com/nurpax/petmate so I just implemented that. Probably makes sense to just merge after any clean up you might have in mind.

twolfson commented on test/gif-encoder_test.js:
```diff
@@ -150,3 +150,42 @@ describe('GifEncoder encoding a multi-framed image with a transparent background
+  for (let y = 0; y < sz; y++) {
+    for (let x = 0; x < sz; x++) {
+      const offs = y*sz + x;
+      const white = Math.floor(y/(sz/2)) ^ Math.floor(x/(sz/2)) ^ 1;
```
Can you explain this line further?

nurpax replied:
This creates a 10x10 checkerboard pattern (5x5 square) of indexed colors 0 and 1. It was easier to generate programmatically than author a .gif for it.

If you run in the debug mode you can see the gif it outputs. 

The xor trick is how one usually generates a 2d checkerboard pattern, in case that’s the unclear part.  

The variable name ”white” is probably not the greatest choice. Maybe just ”colorIndex” would be better.

nurpax replied:
This creates a 10x10 checkerboard pattern (5x5 square) of indexed colors 0 and 1. It was easier to generate programmatically than author a .gif for it.

If you run in the debug mode you can see the gif it outputs. 

The xor trick is how one usually generates a 2d checkerboard pattern, in case that’s the unclear part.  

The variable name ”white” is probably not the greatest choice. Maybe just ”colorIndex” would be better.

twolfson replied:
Ah, I see. It might be even better to enumerate the index values as an array literal -- that would avoid future confusion as well:

```js
var indexedPixels = [
  0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
  1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
  0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
  1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
  0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
  1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
  0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
  1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
  0, 1, 0, 1, 0, 1, 0, 1, 0, 1,
  1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
];
```

twolfson reviewed (changes requested):
Ooooh, I see. Yea, prob want to support both via the API. I'm not sure people can provide indexed pixels so easily. Maybe add a new `option` to `writeFrame` or on `initialization` which is `treatPixelsAsIndexedPixels`

nurpax commented:
Maybe gifencoder options could have a pixelFormat field which could be either RGBA, RGB (can’t recall which you have now) or indexed. This would be explicit and similar to what other imaging apis might have.

twolfson commented:
I'd prefer to stick to a consistent RGBA input format to avoid changing the API too much. I'm more considering this repo to be in stable API mode than anything else due to my lack of time

I think I'm realizing that the custom palette lookup might be a pain to implement so I'll take some more time to think about it. I'll reply again by the end of next week

twolfson commented:
Sorry for the delay. I've sided with keeping the PR mostly as is but making the API flexible for option (2) in the future

This should look like:

```js
var gif = new GIF({palette: ...});
gif.addFrame({
  indexedPixels: true
}, [0, 1, ...]);
```

No need to update anything on your end. I'm going to implement these changes on top of the PR to minimize future back/forths

twolfson commented:
Okay, done with the updates and released it in `0.7.0`

As a heads up, the calculation for `palSize` was incorrect -- it's always 7 due to using an 8-bit color table. `GIFEncoder.js` was correcting this inside of `analyzePixels`:

I also adjusted the palette code to properly represent what the software is currently doing which is writing a new palette every frame

```js
var gif = new GIF(10, 10);
var palette = [0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF];
gif.addFrame([
  0, 1, ...
], {
  palette: palette,
  indexedPixels: true
});
```

I'm happy to accept a PR which makes it not do that but I felt it was lying through its teeth

Thanks for the PR and quick back/forths!
